File: Assignment 1/task3.py
import sys
import json
import time
from pyspark import SparkContext

def process_task3(review_path, business_path, output_path_a, output_path_b):
    """
    Computes average stars per city and compares sorting methods.
    """

    # Step 1: Initialize Spark Context
    sc = SparkContext("local[*]", "DSCI 553 Task 3")
    sc.setLogLevel("ERROR")

    try:
        # Step 2: Load datasets and parse JSON records
        review_rdd = sc.textFile(review_path).map(json.loads)
        business_rdd = sc.textFile(business_path).map(json.loads)

        # Step 3: Extract relevant fields
        review_stars_rdd = review_rdd.map(lambda x: (x["business_id"], x["stars"]))
        business_city_rdd = business_rdd.map(lambda x: (x["business_id"], x["city"]))

        # Step 4: Join datasets
        joined_rdd = review_stars_rdd.join(business_city_rdd)

        # Step 5: Compute average stars per city
        avg_stars_per_city = (
            joined_rdd.map(lambda x: (x[1][1], (x[1][0], 1)))  # (city, (stars, count))
            .reduceByKey(lambda a, b: (a[0] + b[0], a[1] + b[1]))  # Aggregate sums
            .mapValues(lambda x: round(x[0] / x[1], 2))  # Compute average
        )

        # Step 6: Sort by stars descending, then city name lexicographically
        sorted_avg_stars = avg_stars_per_city.sortBy(lambda x: (-x[1], x[0]))

        # Step 7: Save output for Task 3(A)
        with open(output_path_a, "w") as f:
            f.write("city,stars\n")
            for city, stars in sorted_avg_stars.collect():
                f.write(f"{city},{stars}\n")

        # Step 8: Compare sorting methods for top 10 cities

        # Method 1: Sort using Python
        start_time = time.time()
        sorted_avg_stars_py = sorted_avg_stars.take(100)  # Avoid collecting everything
        sorted_avg_stars_py = sorted(sorted_avg_stars_py, key=lambda x: (-x[1], x[0]))[:10]
        m1_time = round(time.time() - start_time, 5)

        # Method 2: Sort using Spark’s takeOrdered
        start_time = time.time()
        sorted_avg_stars_spark = sorted_avg_stars.takeOrdered(10, key=lambda x: (-x[1], x[0]))
        m2_time = round(time.time() - start_time, 5)

        # Step 9: Save output for Task 3(B)
        output_b = {
            "m1": m1_time,
            "m2": m2_time,
            "reason": "Spark sorting is optimized for distributed computation, but since the dataset is small, Python sorting performs similarly."
        }

        with open(output_path_b, "w") as f:
            json.dump(output_b, f, indent=4)

    finally:
        sc.stop()

if __name__ == "__main__":
    if len(sys.argv) != 5:
        print("Usage: spark-submit task3.py <review_file> <business_file> <output_file_a> <output_file_b>")
        sys.exit(1)

    process_task3(sys.argv[1], sys.argv[2], sys.argv[3], sys.argv[4])






























# import sys
# import json
# import time
# from pyspark import SparkContext


# def process_task3(review_path, business_path, output_path_a, output_path_b):
#     """
#     Process Yelp reviews and business data to compute the average stars per city and compare sorting methods.
#     """
#     # Initialize SparkContext
#     sc = SparkContext("local[*]", "DSCI 553 Task 3")
#     sc.setLogLevel("ERROR")

#     try:
#         # Load datasets and convert to RDDs
#         review_rdd = sc.textFile(review_path).map(json.loads)
#         business_rdd = sc.textFile(business_path).map(json.loads)

#         # Extract relevant fields
#         review_stars_rdd = review_rdd.map(lambda x: (x["business_id"], x["stars"]))
#         business_city_rdd = business_rdd.map(lambda x: (x["business_id"], x["city"]))

#         # Join datasets on business_id
#         joined_rdd = review_stars_rdd.join(business_city_rdd)

#         # Compute average stars per city
#         avg_stars_per_city = (
#             joined_rdd.map(lambda x: (x[1][1], (x[1][0], 1)))  # (city, (stars, count))
#             .reduceByKey(lambda a, b: (a[0] + b[0], a[1] + b[1]))  # Aggregate sums
#             .mapValues(lambda x: round(x[0] / x[1], 2))  # Compute average with rounding
#         )

#         # Sort by stars descending, then city name lexicographically
#         sorted_avg_stars = avg_stars_per_city.sortBy(lambda x: (-x[1], x[0]))

#         # Save output for Task 3(A)
#         with open(output_path_a, "w") as f:
#             f.write("city,stars\n")
#             for city, stars in sorted_avg_stars.collect():
#                 f.write(f"{city},{stars}\n")

#         # Compare sorting methods for top 10 cities

#         # Method 1: Sort using Python
#         start_time = time.time()
#         sorted_avg_stars_py = sorted_avg_stars.collect()
#         sorted_avg_stars_py = sorted(sorted_avg_stars_py, key=lambda x: (-x[1], x[0]))[:10]
#         m1_time = round(time.time() - start_time, 5)

#         # Method 2: Sort using Spark's takeOrdered
#         start_time = time.time()
#         sorted_avg_stars_spark = sorted_avg_stars.takeOrdered(10, key=lambda x: (-x[1], x[0]))
#         m2_time = round(time.time() - start_time, 5)

#         # Save output for Task 3(B)
#         output_b = {
#             "m1": m1_time,
#             "m2": m2_time,
#             "reason": "Spark sorting is optimized for distributed computation, but in this case, the dataset is small, so Python sorting performs similarly."
#         }

#         with open(output_path_b, "w") as f:
#             json.dump(output_b, f, indent=4)

#         print(f"✅ Task 3 completed. Output saved to {output_path_a} and {output_path_b}")

#     except Exception as e:
#         print(f"❌ ERROR: {str(e)}")
#         sys.exit(1)

#     finally:
#         sc.stop()


# if __name__ == "__main__":
#     if len(sys.argv) != 5:
#         print("Usage: spark-submit task3.py <review_file> <business_file> <output_file_a> <output_file_b>")
#         sys.exit(1)

#     process_task3(sys.argv[1], sys.argv[2], sys.argv[3], sys.argv[4])

File: Assignment 1/task3.scala
import org.apache.spark.{SparkConf, SparkContext}
import org.json4s._
import org.json4s.jackson.JsonMethods._
import java.io.{File, PrintWriter}

object task3 {
  implicit val formats: DefaultFormats.type = DefaultFormats

  def main(args: Array[String]): Unit = {
    /*
      Usage:
      spark-submit --class Task3 <jar_file> <review_file> <business_file> <output_file_a> <output_file_b>
    */
    if (args.length != 4) {
      println("Usage: spark-submit --class Task3 <jar_file> <review_file> <business_file> <output_file_a> <output_file_b>")
      sys.exit(1)
    }

    val reviewPath   = args(0)
    val businessPath = args(1)
    val outputA      = args(2)
    val outputB      = args(3)

    val conf = new SparkConf().setAppName("DSCI 553 Task3").setMaster("local[*]")
    val sc   = new SparkContext(conf)
    sc.setLogLevel("ERROR")

    try {
      // 1) Load & parse reviews
      val reviewRDD = sc.textFile(reviewPath).flatMap { line =>
        parseOpt(line)
      }.map { j =>
        val bid   = (j \ "business_id").extract[String]
        val stars = (j \ "stars").extract[Double]
        (bid, stars)
      }

      // 2) Load & parse businesses
      val businessRDD = sc.textFile(businessPath).flatMap { line =>
        parseOpt(line)
      }.map { j =>
        val bid  = (j \ "business_id").extract[String]
        val city = (j \ "city").extract[String]
        (bid, city)
      }

      // 3) Join on business_id => (bid, (stars, city))
      val joinedRDD = reviewRDD.join(businessRDD)
      // => (bid, (stars, city))

      // 4) city => (sumStars, count), then compute average
      val cityStarsRDD = joinedRDD.map {
        case (_, (stars, city)) =>
          (city, (stars, 1L))
      }.reduceByKey {
        case ((s1, c1), (s2, c2)) => (s1 + s2, c1 + c2)
      }.mapValues {
        case (s, c) =>
          // If you need no rounding, just do s/c
          // Here we match PySpark's 2-decimal rounding
          val avg = s / c
          BigDecimal(avg).setScale(2, BigDecimal.RoundingMode.HALF_UP).toDouble
      }

      // 5) Sort by descending average, tie-break ascending city
      //    This matches .sortBy(lambda x: (-x[1], x[0])) in PySpark
      val sortedCityStars = cityStarsRDD.sortBy(
        { case (city, avg) => (-avg, city) },
        ascending = true
      )

      // Collect for writing
      val finalCityStars = sortedCityStars.collect()

      // 6) Write file for Task3 (A)
      val writerA = new PrintWriter(new File(outputA))
      writerA.println("city,stars")
      finalCityStars.foreach {
        case (city, avg) => writerA.println(s"$city,$avg")
      }
      writerA.close()

      // 7) Compare M1 vs. M2 (top 10)
      //    For M1, we measure after we have the (city -> avg) RDD
      val m1Start = System.nanoTime()
      val subset  = finalCityStars.take(100) // local subset
      // sort in Python-like manner: first by -avg, then city
      val top10M1 = subset.sortBy { case (city, avg) => (-avg, city) }.take(10)
      val m1Time  = (System.nanoTime() - m1Start) / 1e9

      // M2: Spark's takeOrdered(10)
      val m2Start = System.nanoTime()
      val top10M2 = sortedCityStars.takeOrdered(10)(
        Ordering.by { case (city, avg) => (-avg, city) }
      )
      val m2Time = (System.nanoTime() - m2Start) / 1e9

      // 8) Write file for Task3 (B)
      val reasonStr = "Spark sorting is optimized for distributed computation, but on small data, Python sorting is also efficient."
      val jsonOutput =
        s"""
           |{
           |  "m1": $m1Time,
           |  "m2": $m2Time,
           |  "reason": "$reasonStr"
           |}
           |""".stripMargin

      val writerB = new PrintWriter(new File(outputB))
      writerB.write(jsonOutput)
      writerB.close()

    } finally {
      sc.stop()
    }
  }
}
